contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,978 @@
+     *             shall be reverted.
+     *             Swap transaction is the one that shall fail. Unfortunately there is no easy way to query Aerodrome
+     *             slipstream pool for the amount of liquidity deposited. Even tokens on the contract can be from
+     *             fees, or liquidity that has been removed but not yet claimed.
+     *             It becomes the responsibility of the strategist or deployer to add some liquidity in the configured
+     *             tick ranges to the pool to facilitate the swap. Effectively turning Scenario 1 into a Scenario 2
+     * Scenario 2: When there is no liquidity in the pool from the strategy but there is from other LPs then
+     *             only step 1 is skipped. (It is important to note that liquidity needs to exist in the configured
+     *             strategy tick ranges in order for the swap to be possible) Step 3 mints new liquidity position
+     *             instead of adding to an existing one.
+     * Scenario 3: When there is strategy's liquidity in the pool all 3 steps are taken
+     *
+     *
+     * Exact _amountToSwap, _minTokenReceived & _swapWeth parameters shall be determined by simulating the
+     * transaction off-chain. The strategy checks that after the swap the share of the tokens is in the
+     * expected ranges.
+     *
+     * @param _amountToSwap The amount of the token to swap
+     * @param _swapWeth Swap using WETH when true, use OETHb when false
+     * @param _minTokenReceived Slippage check -> minimum amount of token expected in return
+     */
+    function rebalance(
+        uint256 _amountToSwap,
+        bool _swapWeth,
+        uint256 _minTokenReceived
+    ) external nonReentrant onlyGovernorOrStrategist {
+        /**
+         * Would be nice to check if there is any total liquidity in the pool before performing this swap
+         * but there is no easy way to do that in UniswapV3:
+         * - clPool.liquidity() -> only liquidity in the active tick
+         * - asset[1&2].balanceOf(address(clPool)) -> will include uncollected tokens of LP providers
+         *   after their liquidity position has been decreased
+         */
+
+        /**
+         * When rebalance is called for the first time there is no strategy's
+         * liquidity in the pool yet. The full liquidity removal is thus skipped.
+         */
+        if (tokenId != 0) {
+            _removeLiquidity(1e18);
+        }
+        // in some cases we will just want to add liquidity and not issue a swap to move the
+        // active trading position within the pool
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _swapWeth, _minTokenReceived);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        _checkForExpectedPoolPrice();
+
+        _addLiquidity();
+        _checkForExpectedPoolPrice();
+    }
```

Comment:
🔴 There's a theoretical way for the strategist to steal a small amount of funds using rebalance. That's okay.

What's not okay is that they can keep doing it infinitely and empty the protocol. What if we did a vault -> token solvency check at the end of rebalance here, that ensure that the vault total value was > 0.998 * oethb total supply? This caps the max that can be stolen in this way.